IdentityCore/src/webview/response/MSIDWebUpgradeRegResponse.m

```diff
@@ -0,0 +1,104 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDWebUpgradeRegResponse.h"
+#import "MSIDWebResponseOperationConstants.h"
+#import "MSIDWebResponseOperationFactory.h"
+#import "MSIDWebResponseBrokerInstallOperation.h"
+#import "MSIDWebWPJResponse+Internal.h"
+
+@implementation MSIDWebUpgradeRegResponse
+
+static NSString *const SCHEME_MSAUTH = @"msauth";
+static NSString *const UPGRADE_REG = @"upgradeReg";
+
++ (void)load
+{
+    [MSIDWebResponseOperationFactory registerOperationClass:MSIDWebResponseBrokerInstallOperation.class forResponseClass:self];
+}
+
+- (instancetype)initWithURL:(NSURL *)url
+                    context:(id<MSIDRequestContext>)context
+                      error:(NSError **)error
+{
+    // Check for upgrade registration
+    if (![self isBrokerUpgradeRegResponse:url])
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDOAuthErrorDomain,
+                                     MSIDErrorServerInvalidResponse,
+                                     [NSString stringWithFormat:
+                                      @"Upgrade registration response should have %@ as a scheme and %@/broker as a host",
+                                        SCHEME_MSAUTH, UPGRADE_REG],
+                                     nil, nil, nil, context.correlationId, nil, NO);
+        }
+        return nil;
+    }
+    
+    return [super initResponseWithURL:url context:context error:error];
+}
+
+/**
+ * return true when the url response is matching a device upgrade registration
+ **/
+- (BOOL)isBrokerUpgradeRegResponse:(NSURL *)url
+{
+    NSString *scheme = url.scheme;
+    NSString *host = url.host;
+    
+    // For embedded webview, if link starts with msauth scheme and contain upgradeReg host
+    // then it is migrateWpj request
+    // e.g. msauth://upgradeReg?param=param
+    if ([scheme isEqualToString:SCHEME_MSAUTH] && [host isEqualToString:UPGRADE_REG]) //$todo upgradeReg
+    {
+        return YES;
+    }
+    
+    NSArray *pathComponents = url.pathComponents;
+    
+    if ([pathComponents count] < 2)
+    {
+        return NO;
+    }
+    
+    // For system webview, this link will start with the redirect uri and will have msauth and upgradeReg as path parameters
+    // e.g. myscheme://auth/msauth/upgradeReg?param=param
+    NSUInteger pathComponentCount = pathComponents.count;
+    
+    if ([pathComponents[pathComponentCount - 1] isEqualToString:UPGRADE_REG]
+        && [pathComponents[pathComponentCount - 2] isEqualToString:SCHEME_MSAUTH])
+    {
+        return YES;
+    }
+    
```

Comment:
I think we want ESTS to also return a tenantId for the registration upgrade rather than the username

Reply:
returning client_info and that is set in the initializer.